### PR TITLE
ARC-2290 new jira client part one

### PR DESCRIPTION
**Pull request by joshkay10**
**What's in this PR?**
Creating a new  JiraClient class to merge the existing ones together, as part of that i have moved helper functions into their own files, this is those helper functions. I also added almost 100% testing coverage for them since i was in the neighbourhood. 

Added feature flags
Not required for this change, no code is being referenced.... yet

### CODE REVIEW

File: src/jira/client/jira-client-deployment-helper.ts [hunk 1]
```diff
@@ -16,3 +18,21 @@ export const getDeploymentDebugInfo = (jiraPayload: JiraDeploymentBulkSubmitData
 		commitCount: associations.filter(a => a.associationType === "commit").map(a => a.values?.length || 0).reduce((a, b) => a + b, 0)
 	};
 };
+
```
Comment (joshkay10) [inline src/jira/client/jira-client-deployment-helper.ts:21]: This code is as was, pulled from jiraclient file and moved here.... Tests are new though

File: src/jira/client/jira-client-issue-key-helper.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
+import { uniq } from "lodash";
```
Comment (joshkay10) [inline src/jira/client/jira-client-issue-key-helper.ts:1]: Most of these are just moved from jiraclient into this file. I will highlight any other changes

File: src/jira/client/jira-client-issue-key-helper.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+ * Truncates branches, commits and PRs to their first ISSUE_KEY_API_LIMIT issue keys
+ */
+export const truncateIssueKeys = (repositoryObj) => {
+	updateRepositoryIssueKeys(repositoryObj, truncate);
+};
+
+/**
+ * Get truncated issue keys and associations based on the ISSUE_KEY_API_LIMIT.
+ */
+export const getTruncatedIssueKeys = (data: IssueKeyObject[] = []): IssueKeyObject[] =>
```
Comment (joshkay10) [inline src/jira/client/jira-client-issue-key-helper.ts:35]: This function was rewritten, during testing it was found that the existing function wasn't truncating keys

File: src/jira/client/jira-client-issue-key-helper.test.ts [hunk 4]
```diff
@@ -0,0 +1,667 @@
+import {
```
Comment (joshkay10) [inline src/jira/client/jira-client-issue-key-helper.test.ts:1]: These are all new :sigh: but ive isolated each function into its own describe block so its pretty easy to follow IMO but will be having a re-review over this too

File: src/jira/client/jira-client-issue-key-helper.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	if (!resources) {
+		return true;
+	}
+
+	const issueKeyCounts = resources.filter(resource => resource.associations?.length > 0).map((resource) => resource.associations[0].values.length);
+	return Math.max(...issueKeyCounts) <= ISSUE_KEY_API_LIMIT;
+};
+
+/**
+ * Extracts unique issue keys and hashes them
```
Comment (joshkay10) [inline src/jira/client/jira-client-issue-key-helper.ts:148]: Hashing issue keys is BS, i want to not hash issue keys!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Will consult UGC/PII rules to see if we can not do this

File: src/jira/client/jira-client-issue-key-helper.ts [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	if (repositoryObj.pullRequests) {
+		repositoryObj.pullRequests = updateIssueKeysFor(repositoryObj.pullRequests, mutatingFunc);
+	}
+};
+
+/**
+ * Finds the first association of type "issueIdOrKeys" in a given resource.
+ */
+export const findIssueKeyAssociation = (resource: IssueKeyObject): JiraAssociation | undefined => {
+	return resource.associations?.find(a => a.associationType == "issueIdOrKeys");
```
Comment (bgvozdev) [inline src/jira/client/jira-client-issue-key-helper.ts:99]: nit `===`

File: src/jira/client/jira-client-issue-key-helper.ts [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+	return resources;
+};
+
+/**
+ * Runs the mutatingFunc on the association values field for each entity resource
+ * Assumption is that the transformed resource only has one association which is for
+ * "issueIdOrKeys" association.
+ */
+export const updateIssueKeyAssociationValuesFor = (resources: JiraRemoteLink[], mutatingFunc: any): JiraRemoteLink[] => {
+	resources?.forEach(resource => {
```
Comment (bgvozdev) [inline src/jira/client/jira-client-issue-key-helper.ts:124]: nit: why `?`

File: src/jira/client/jira-client-issue-key-helper.ts [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+		const association = findIssueKeyAssociation(resource);
+		if (association) {
+			association.values = mutatingFunc(resource.associations[0].values);
```
Comment (bgvozdev) [inline src/jira/client/jira-client-issue-key-helper.ts:127]: Could you please explain the logic? Why are we taking the very first association fro the list when there's one issueKeyAssociation exists?

File: src/jira/client/jira-client-issue-key-helper.ts [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+	return resources;
+};
+
+/**
+ * Returns if the max length of the issue key field is within the limit
+ * Assumption is that the transformed resource only has one association which is for
+ * "issueIdOrKeys" association.
+ */
+export const withinIssueKeyAssociationsLimit = (resources: JiraRemoteLink[]): boolean => {
+	if (!resources) {
```
Comment (bgvozdev) [inline src/jira/client/jira-client-issue-key-helper.ts:139]: nit `if (resource.length === 0)` ?

File: src/jira/client/jira-client-issue-key-helper.ts [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+		return true;
+	}
+
+	const issueKeyCounts = resources.filter(resource => resource.associations?.length > 0).map((resource) => resource.associations[0].values.length);
```
Comment (bgvozdev) [inline src/jira/client/jira-client-issue-key-helper.ts:143]: Could you please explain the logic, why do we pick the very first association if there are any?

Comment (joshkay10) [reply src/jira/client/jira-client-issue-key-helper.ts:143]: This is copied from jiraclients file and landed here to tidy up. no new logic. :excite:

File: src/jira/client/jira-client-issue-key-helper.ts (hunk 9 above)
Comment (joshkay10) [reply src/jira/client/jira-client-issue-key-helper.ts:139]: This is copied from jiraclients file and landed here to tidy up. no new logic. :excite:

File: src/jira/client/jira-client-issue-key-helper.ts (hunk 8 above)
Comment (joshkay10) [reply src/jira/client/jira-client-issue-key-helper.ts:127]: This is copied from jiraclients file and landed here to tidy up. no new logic. :excite:

File: src/jira/client/jira-client-issue-key-helper.ts (hunk 7 above)
Comment (joshkay10) [reply src/jira/client/jira-client-issue-key-helper.ts:124]: This is copied from jiraclients file and landed here to tidy up. no new logic. :excite:

File: src/jira/client/jira-client-issue-key-helper.ts (hunk 6 above)
Comment (joshkay10) [reply src/jira/client/jira-client-issue-key-helper.ts:99]: This is copied from jiraclients file and landed here to tidy up. no new logic. :excite:

**Review by bgvozdev — APPROVED** [review]

**Pull request merged**